lib/appsignal/hooks/puma.rb

```diff
@@ -33,5 +37,57 @@ def stop_workers
         end
       end
     end
+
+    class PumaProbe
+      def initialize
+        @hostname = Appsignal.config[:hostname] || Socket.gethostname
+      end
+
+      def call
+        return unless ::Puma.stats
+
+        stats = JSON.parse Puma.stats, :symbolize_names => true
+        counts = {
+          :backlog => 0,
+          :running => 0,
+          :pool_capacity => 0,
+          :max_threads => 0
+        }
+
+        if stats[:worker_status] # Multiple workers
+          stats[:worker_status].each do |worker|
+            stat = worker[:last_status]
+
+            counts[:backlog] += stat[:backlog]
+            counts[:running] += stat[:running]
+            counts[:pool_capacity] += stat[:pool_capacity]
+            counts[:max_threads] += stat[:max_threads]
+          end
+
+          gauge(:workers, stats[:workers], :kind => :count)
+          gauge(:workers, stats[:booted_workers], :kind => :booted)
+          gauge(:workers, stats[:old_workers], :kind => :old)
+
+        else # Single worker
+          counts[:backlog] += stats[:backlog]
+          counts[:running] += stats[:running]
+          counts[:pool_capacity] += stats[:pool_capacity]
+          counts[:max_threads] += stats[:max_threads]
+        end
+
+        gauge(:backlog, counts[:backlog])
+        gauge(:running, counts[:running])
```

Comment:
What does this report? whether or not it's running or not?

Reply:
The number of running workers/processes

Reply:
The naming is a bit vague still. Can we rename it to `threads_running`? (since puma uses threads.) It would be `threads * workers = threads_running`

or can we group it with `max_threads` and tag it?

```ruby
gauge "threads", counts[:running], :type => :running
gauge "threads", counts[:max_threads], :type => :max
```

Reply:
If we name the counter `threads`, I think `max`/`running` should work?

Reply:
I've updated the yaml to reflect the changes, let me know if that's the way to go and I'll update the code :)

Reply:
I hope so, yes 😄 
They seem to be about the same metric, just one is the value limit and the other the current value

Reply:
> I've updated the yaml to reflect the changes, let me know if that's the way to go and I'll update the code :)

@matsimitsu The YAML is missing the `puma_pool_capacity` metric and I have some thoughts maybe on the line label format, but that's another PR

Reply:
Check, `puma_pool_capacity` is the "negative backpressure" ( https://github.com/puma/puma/pull/1579 ) and a bit hard to explain, I think. (maybe we should have comments per graph ;))

Reply:
it is yes. But some metric are a bit different. Some metrics spike and some drop when it's cause for worry.

Do you mean you would be against showing it? I think it's a valid metric to show, albeit with some extra description in the docs.

> maybe we should have comments per graph 

It's not a crazy idea to add that

Reply:
> Do you mean you would be against showing it? I think it's a valid metric to show, albeit with some extra description in the docs.

I would be if there's no way to explain what it is and it's implication are. If we're adding `description` fields to individual graphs, I'm for adding it :) 

(we should make sure they don't show as `null` in the editor when empty)

Reply:
Not sure if that should be a blocker for this first release. Would require some input from Wes for a way to show it in the UI itself. So if it is a blocker, let's remove it for now and add it in a future version of the dashboard?

> (we should make sure they don't show as null in the editor when empty)
let's fix that on the server for the dashboard description at least. Let's create an issue for it.

Reply:
I'll pick it up in https://github.com/appsignal/appsignal-server/issues/4135

@matsimitsu But per graph descriptions could be an issue without having to send it in the GraphQL query? (Maybe move this discussion to a server issue?)

Reply:
Updated the code/yaml

Reply:
ok so you want to add the `puma_connection_backlog` metric for now?

And add the graph descriptions (https://github.com/appsignal/appsignal-server/issues/4137) later?

Reply:
Yes, it's a nice metric to add triggers on :)

Reply:
ok! Resolving this conversation then :)